### PR TITLE
[ENH] Add predict_quantiles to ets, pmdarima adapter

### PR DESCRIPTION
**Pull request by kejsitake**
#### Reference Issues/PRs

Progress on #1687 

#### What does this implement/fix? Explain your changes.

- [x] Implements predict_quantiles for ets
- [x] Implements predict_quantiles for pmdarima adapter
- [x] extends pmdarima to support a list of alphas instead of a single alpha

### CODE REVIEW

**Comment by fkiraly** [comment]
looks good on superficial view.

May I ask, what was the outcome of your investigation that ETS produced non-symmetric predictive intervals?
(non-symmetric around the mean/point forecast)

**Comment by kejsitake** [comment]
I had misunderstood the output of predict from statsmodels. Once I printed that out, it became clearer and I could spot the mistake. This issue has been solved on this version.

For consistency with tbats and theta, the output of predict-interval is still a two level dataframe, but I can change that if necessary.

**Comment by fkiraly** [comment]
> For consistency with tbats and theta, the output of predict-interval is still a two level dataframe, but I can change that if necessary.

Yes, fully agreed - let's keep it that way, for now. Let's not mix output types, and change it either not at all or consistently.

File: sktime/forecasting/ets.py [hunk 1]
```diff
@@ -447,15 +404,64 @@ def _predict(
...
+            Row index is fh. Entries are quantile forecasts, for var in col index,
+                at quantile probability in second col index, for the row index.
+        """
+        self.check_is_fitted()
+        alpha = check_alpha(alpha)
+        # convert alpha to the one needed for predict intervals
+        alphas = []
+        for a in alpha:
+            if a < 0.5:
+                alphas.append(round(((0.5 - a) * 2), 4))
```
Comment (fkiraly) [inline sktime/forecasting/ets.py:440]: I think no rounding should take place here

File: sktime/forecasting/ets.py [hunk 2]
```diff
@@ -447,15 +404,64 @@ def _predict(
...
+        valid_indices = fh.to_absolute(self.cutoff).to_pandas()
+
+        start, end = valid_indices[[0, -1]]
+        prediction_results = self._fitted_forecaster.get_prediction(
+            start=start, end=end
+        )
+
+        pred_quantiles = pd.DataFrame()
+        for a, coverage in zip(alpha, alphas):
+            pred_int = prediction_results.pred_int(round(1 - coverage, 4))
```
Comment (fkiraly) [inline sktime/forecasting/ets.py:455]: or here

File: sktime/forecasting/ets.py [hunk 3]
```diff
@@ -447,15 +404,64 @@ def _predict(
...
+            pred_int.columns = ["lower", "upper"]
+            if a < 0.5:
+                pred_quantiles[a] = pred_int["lower"].loc[valid_indices]
+            else:
+                pred_quantiles[a] = pred_int["upper"].loc[valid_indices]
+        arrays = [len(alpha) * ["Quantiles"], alpha]
+        index = pd.MultiIndex.from_tuples(list(zip(*arrays)))
```
Comment (fkiraly) [inline sktime/forecasting/ets.py:462]: if you use `from_prod`, zou don't need to use `zip` etc.

File: sktime/forecasting/base/adapters/_pmdarima.py [hunk 4]
```diff
@@ -117,45 +131,162 @@ def _predict_in_sample(
...
+
+        Returns
+        -------
+        quantiles : pd.DataFrame
+            Column has multi-index: first level is variable name from y in fit,
+                second level being the values of alpha passed to the function.
+            Row index is fh. Entries are quantile forecasts, for var in col index,
+                at quantile probability in second col index, for the row index.
+        """
+        self.check_is_fitted()
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pmdarima.py:228]: this line is unnecessary, already checked in the public method

File: sktime/forecasting/base/adapters/_pmdarima.py [hunk 5]
```diff
@@ -117,45 +131,162 @@ def _predict_in_sample(
...
+        alpha = check_alpha(alpha)
+
+        # convert alpha to the one needed for predict intervals
+        alphas = []
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pmdarima.py:232]: perhaps better call this `coverage`

File: sktime/forecasting/base/adapters/_pmdarima.py [hunk 6]
```diff
@@ -117,45 +131,162 @@ def _predict_in_sample(
...
+        if self.fh.is_all_out_of_sample(self.cutoff):
+            y_oos_pred, y_oos_pred_int = self._predict_fixed_cutoff(fh_oos, **kwargs)
+            pred_quantiles = pd.DataFrame()
+            for oos_intervals, a in zip(y_oos_pred_int, alpha):
+                if a < 0.5:
+                    pred_quantiles[a] = oos_intervals["lower"]
+                else:
+                    pred_quantiles[a] = oos_intervals["upper"]
+            arrays = [len(alpha) * ["Quantiles"], alpha]
+            index = pd.MultiIndex.from_tuples(list(zip(*arrays)))
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pmdarima.py:255]: if you use `from_prod`, no need to use `zip`

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice! This is the last existing probabilistic forecaster, right?

Some minor changes above.

The one that I'd consider is a blocker is the rounding of `alpha` in your `_predict_quantiles`. This is because this violates the interface contract - the floats in the column `MultiIndex` should be identical with the `alpha` passed to `_predict_quantiles`.

File: sktime/forecasting/base/adapters/_pmdarima.py (hunk 4 above)
Comment (kejsitake) [reply sktime/forecasting/base/adapters/_pmdarima.py:228]: Removed

File: sktime/forecasting/base/adapters/_pmdarima.py (hunk 5 above)
Comment (kejsitake) [reply sktime/forecasting/base/adapters/_pmdarima.py]: changed in the new verison

File: sktime/forecasting/base/adapters/_pmdarima.py (hunk 6 above)
Comment (kejsitake) [reply sktime/forecasting/base/adapters/_pmdarima.py]: edited

**Comment by kejsitake** [comment]
Thanks for the comments! I believe I fixed all of the issues! Yes, I think these are the last two probabilistic forecasters.

**Comment by fkiraly** [comment]
are you certain you have pushed your newest changes to the remote branch?
It does appear there have been no changes since my review, so I suspect you may have accidentally forgotten to do this.

**Comment by kejsitake** [comment]
Hi! I had amended my commit, I don't know if that makes it more difficult to review or was causing the issue, but I can avoid doing this in the future. I also spotted some other small things, so I added a new commit.

In the 4th tab `Files changed`, I can see that `predict_quantiles` doesn't round the alpha anymore and the usage of `pd.multiIndex.from_product`, which was not there in the previous version.

**Review by fkiraly — APPROVED** [review]
Looks good now, thanks!

Maybe it doesn´t properly update the displayed changes if you amend a commit, instead of making a new commit?
Anyway, thanks, all fine now.

**Pull request merged**